### PR TITLE
#2559. Add constructor augmentation tests. Part 7. Redirecting generative constructors

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

**Review by eernstg — COMMENTED** [review]
Looks very good!

I'm suggesting one extra test (could be in this PR or another one).

One thing calls for an adjustment: LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t04_lib.dart should expect that it is not an error to replace one generative redirection by another one.

File: LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t05_lib.dart [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// SharedOptions=--enable-experiment=macros
+
+augment library 'augmenting_constructors_A18_t05.dart';
+
+augment class C {
+  augment C(): this();
+//             ^
+// [analyzer] unspecified
+// [cfe] unspecified
+  augment C.foo(): this.foo();
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t05_lib.dart:23]: Added `augmenting_constructors_A18_t07.dart`. BTW there is no such statement in the spec that cyclic constructor is an error. The same about default values for an optional parameters. May be it makes sense to add it to the spec?

**Review by sgrekhov — COMMENTED** [review]
Updated. PTAL.

File: LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t04_lib.dart [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+/// redirecting generative constructor if the augmented constructor is already
+/// redirecting.
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=macros
+
+augment library 'augmenting_constructors_A18_t04.dart';
+
+augment class C {
+  augment C.bar(): this.foo();
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t04_lib.dart:20]: OK, I don't think we should consider `: this();` to be an initializer list. For instance, it cannot become `: this(), super();`, neither implicitly nor explicitly. So the rules about initializer lists should not be used with generative redirections like `this()`.

File: LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t05_lib.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t05_lib.dart:23]: Thanks!

File: LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t04_lib.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Augmentation-libraries/augmenting_constructors_A18_t04_lib.dart:20]: With that in mind, I don't think this test has an error.

**Review by eernstg — COMMENTED** [review]
I think one test expects an error, but shouldn't.

**Review by eernstg — APPROVED** [review]
LGTM!

**Pull request merged**